Configurable page numbers in PDFs

carlobeltrame opened the pull request.

carlobeltrame committed:
Implement configurable page numbers in PDFs

carlobeltrame committed:
Add page numbers to client-print TOC

BacLuc commented on frontend/src/components/print/__tests__/repairPrintConfig.spec.js:
```diff
@@ -136,6 +136,7 @@ describe('repairConfig', () => {
         },
       ],
       documentName: 'test camp',
+      options: { pageNumbers: false },
```
maybe this test could use a little drying in a future pr.

BacLuc commented on print/server/api/pdf.js:
```diff
@@ -121,14 +121,18 @@ export default defineEventHandler(async (event) => {
     })
     measurePerformance(performanceMeasurements, 'load_content')
 
+    const footerTemplate = JSON.parse(queryParams.config).options?.pageNumbers
```
Can i have that in a factory function and not in a ternary statement?

carlobeltrame replied:
I can change it, but I'm not sure I understand how you'd like it to work. Do you mean wrapping this code in a function and using if-else instead of a ternary expression? Or an object mapping two keys to the two static strings? (I think an object with `true` and `false` as keys is not possible, so I'll have to use a ternary or if-else either way then to calculate the correct object key...)

BacLuc replied:
if (JSON.parse(queryParams.config).options?.pageNumbers)
 ....
else
....

you can do this with a let or with a function with 2 returns, how you like.

or
const footerWithPageNumber = ...
const footerNoPageNumber = ....
(JSON.parse(queryParams.config).options?.pageNumbers) ? footerWithPageNumber : footerNoPageNumber;

But the naming here gets difficult for the third option that needs to be supported.

carlobeltrame replied:
I tried an improvement in 7e10a79. Let me know if this improves the situation for you.

BacLuc replied:
:heart:  thank you

BacLuc commented:
@coderabbitai review this pr

carlobeltrame commented on frontend/src/locales/de.json:
```diff
@@ -403,7 +403,9 @@
           "Toc": "Inhaltsverzeichnis"
-        }
+        },
+        "options": "Einstellungen",
+        "pageNumbers": "Seitenzahlen"
       },
```
Actually, in this project the translations are done via Crowdin, and it's better to leave out the translations. Crowdin PRs are otherwise confused and sometimes try to revert changes which we deliberately made via PRs. We also have a fallback system, so the UI does not break even when some of the translations are missing in some languages.

carlobeltrame replied:
Hm actually that's not quite right. The fallback system needs the english translations. I'll add those, but deliberately leave out the french, italian and rumantsch ones.

carlobeltrame replied:
Fixed in 5734bfb

carlobeltrame committed:
Add english translations for fallback

carlobeltrame committed:
Never print the page numbers on the cover page

carlobeltrame committed:
Make sure not to accidentally render a page number

Also, extract the logic as a separate component with a more speaking
name and explaining comment.

carlobeltrame commented on pdf/src/campPrint/tableOfContents/TocEntryPageNumber.vue:
```diff
@@ -0,0 +1,23 @@
+  extends: PdfComponent,
+}
+</script>
+<pdf-style>
+.toc-dotted-line {
+  border-bottom: 1px dotted black;
+  flex-grow: 1;
+  margin-left: 6pt;
+}
+.toc-entry-page-number {
+  width: 24pt;
+  text-align: right;
+}
```
The dotted line is supposed to be on the text baseline (as is kind of usual in tables of contents), so this is actually already more or less the way it's supposed to be.

<img width="947" height="202" alt="image" src="https://github.com/user-attachments/assets/6f6a0dad-1bdd-43c0-926c-7ce335d2a879" />

carlobeltrame committed:
Make sure not to render undefined

carlobeltrame commented on pdf/src/CampPrint.vue:
```diff
@@ -7,7 +7,14 @@
         :config="config"
         :content="content"
-      ></component>
+      >
+        <Text
+          v-if="config.options.pageNumbers"
+          :render="({ pageNumber, totalPages }) => `${pageNumber} / ${totalPages}`"
+          fixed
+          class="page-number"
+        />
+      </component>
```
This on first sight looked like a nice catch, but it's wrong. The SafetyConsiderations.vue actually already contains the default slot. CodeRabbit confused the SafetyConsiderations summary pdf component with the safety considerations content node. The former needs (and already has) the default slot, the latter does not need it, because it does not contain a `<Page>` element.

I did implement the disabled page numbers on the cover page, but I opted to do it by omitting the default slot from the Cover component instead. Fixed in add823e

carlobeltrame commented on pdf/src/campPrint/picasso/PicassoPage.vue:
```diff
@@ -1,5 +1,7 @@
 <template>
   <Page size="A4" :orientation="orientation" class="page">
+    <slot></slot>
+    <View :render="({ pageNumber }) => ($toc[`${id}-${period.id}`] = pageNumber)" />
```
This one is a nice catch. I extracted these "renderless" pieces into a component which is hopefully also easier to understand in the future. Fixed in 71078b5

carlobeltrame commented on pdf/src/campPrint/tableOfContents/TocEntryPageNumber.vue:
```diff
@@ -0,0 +1,23 @@
+<template>
+  <View class="toc-dotted-line" />
+  <Text :render="() => $toc[id]" class="toc-entry-page-number" />
+</template>
```
Fixed in 0840055

carlobeltrame committed:
Try to improve code legibility
